pennylane/fourier/coefficients.py

```diff
@@ -196,43 +214,49 @@ def circuit(inpt):
     return coeffs
 
 
-def _coefficients_no_filter(f, n_inputs, degree):
+def _coefficients_no_filter(f, degree, use_broadcasting):
     r"""Computes the first :math:`2d+1` Fourier coefficients of a :math:`2\pi` periodic
     function, where :math:`d` is the highest desired frequency in the Fourier spectrum.
 
     This function computes the coefficients blindly without any filtering applied, and
     is thus used as a helper function for the true ``coefficients`` function.
 
     Args:
-        f (callable): function that takes a 1D array of ``n_inputs`` scalar inputs
-        n_inputs (int): number of function inputs
-        degree (int): max frequency of Fourier coeffs to be computed. For degree :math:`d`,
-            the coefficients from frequencies :math:`-d, -d+1,...0,..., d-1, d ` will be computed.
+        f (callable): function that takes a 1D array of scalar inputs
+        degree (int or tuple[int]): max frequency of Fourier coeffs to be computed. For degree
+            :math:`d`, the coefficients from frequencies :math:`-d, -d+1,...0,..., d-1, d`
+            will be computed.
+        use_broadcasting (bool): Whether or not to broadcast the parameters to execute
+            multiple function calls at once. Broadcasting is performed along the last axis
+            of the grid of evaluation points.
 
     Returns:
         array[complex]: The Fourier coefficients of the function f up to the specified degree.
     """
+    degree = np.array(degree)
 
-    # number of integer values for the indices n_i = -degree,...,0,...,degree
+    # number of integer values for the indices n_i = -degree_i,...,0,...,degree_i
     k = 2 * degree + 1
 
-    # create generator for indices nvec = (n1, ..., nN), ranging from (-d,...,-d) to (d,...,d).
-    n_range = np.array(range(-degree, degree + 1))
-    n_ranges = [n_range] * n_inputs
-    nvecs = product(*n_ranges)
+    # create generator for indices nvec = (n1, ..., nN), ranging from (-d1,...,-dN) to (d1,...,dN)
+    n_ranges = [np.array(range(-d, d + 1)) for d in degree]
+    nvecs = product(*(n_ranges[:-1] if use_broadcasting else n_ranges))
 
     # here we will collect the discretized values of function f
-    shp = tuple([k] * n_inputs)
-    f_discrete = np.zeros(shape=shp)
-
-    for nvec in nvecs:
-        nvec = np.array(nvec)
+    f_discrete = np.zeros(shape=tuple(k))
 
-        # compute the evaluation points for frequencies nvec
-        sample_points = 2 * np.pi / k * nvec
+    spacing = (2 * np.pi) / k
 
+    for nvec in nvecs:
+        # As we fill in an array into the last entry of an array,
+        # this is a ragged array when use_broadcasting=True
```

Comment:
can you add a test for this? I'd rather this weren't the case at all, but if you feel it's necessary I wouldn't mind a nice example to demonstrate why

Reply:
This is being tested already, and the reason for the if-else clause:
When `use_broadcasting=True`, `nvec` is a `tuple` like `(-2, -1, [-3, -2, -1, 0, 1, 2, 3])`, which is a valid numpy index.
The `else` clause (which is the previous behaviour) deals with `nvec` by casting it to an array and multiplying with the `ndarray` `spacing`. The above `tuple` would yield a ragged array, requiring us to do the multiplication via manual iteration over the first axis of `nvec` and `spacing`.

Reply:
I rewrote the comment to reflect what is actually happening here :)